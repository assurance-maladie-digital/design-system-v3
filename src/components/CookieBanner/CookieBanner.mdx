import {Controls, Canvas, Meta, Source} from '@storybook/blocks';
import * as CookieBannerStories from './CookieBanner.stories';

<Meta of={CookieBannerStories} />

# CookieBanner

Le composant `CookieBanner` est utilisé pour afficher une bannière de consentement pour les cookies.

<Canvas of={CookieBannerStories.Default} />

# API

<Controls of={CookieBannerStories.Default} />

## Sans selection des cookies

Vous pouvez omettre la props `items`, dans ce cas la l'interface de sélection des categories de cookies ne sera pas affichée.
Vous pourrez toujours afficher utiliser l’événement `customize` pour afficher une interface de sélection des cookies personnalisée.

<Canvas
	of={CookieBannerStories.WithoutCookiesItems}
	source={{
		language: 'html',
		format: 'dedent',
		code: `
		<script lang="ts" setup>
			import { CookieBanner } from '@cnamts/synapse'

			const openBanner = ref(true)
		</script>

		<template>
			<CookieBanner
				v-model="openBanner"
				@customize="openDialog = true"
			/>
		</template>
		`
	}}
/>

## Description de la bannière

Vous pouvez personnaliser le texte de la bannière en utilisant le slot par défaut

<Canvas
	of={CookieBannerStories.DescriptionSlot}
	source={{
		language: 'html',
		format: 'dedent',
		code: `
		<script lang="ts" setup>
			import { CookieBanner } from '@cnamts/synapse'

			const openBanner = ref(true)
		</script>

		<template>
			<CookieBanner
				v-model="openBanner"
				@customize="openDialog = true"
			>
				<p><b>Custom</b> description</p>
			</CookieBanner>
		</template>
		`
	}}
/>


## Personnalisation des composants Vuetify

Vous pouvez personnaliser les composants Vuetify utilisés par le composant `CookieBanner` en utilisant la props `vuetifyOptions`.


<Canvas
	of={CookieBannerStories.Customization}
	source={{
		language: 'html',
		format: 'dedent',
		code: `
		<script lang="ts" setup>
			import { CookieBanner } from '@cnamts/synapse'

			const openBanner = ref(true)

			const items = {
				essentials: [
					{
						name: 'session',
						description: 'Sauvegarde la session pour rester connecté.',
						conservation: '20 heures',
					},
					{
						name: 'cookie_policy',
						description: 'Sauvegarde les préférences de cookies.',
						conservation: '1 an',
					},
				],
				functional: [
					{
						name: 'contrast',
						description: 'Sauvegarde la personnalisation de l’affichage.',
						conservation: '1 an',
					},
				],
				analytics: [
					{
						name: 'user_id',
						description: 'Sauvegarde l’identifiant unique de visiteur.',
						conservation: '6 mois',
					},
				],
			}

			const options = {
				banner: {
					color: '#ced9eb',
				},
				customizeBtn: {
					variant: 'text',
					color: 'orange',
				},
				rejectBtn: {
					variant: 'outlined',
				},
				acceptBtn: {
					variant: 'outlined',
				},
			}

			function onSubmit(e) {
				console.log('Les cookies suivants ont été sélectionnés :', e)
				alert('Vos préférences ont été enregistrées.')

				// Charger les cookies non essentiels
			}
		</script>

		<template>
			<CookieBanner
				v-model="openBanner"
				:vuetifyOptions="options"
			/>
		</template>
		`
	}}
/>