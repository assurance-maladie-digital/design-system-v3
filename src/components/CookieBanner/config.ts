const commonBtnConfig = {
	color: 'primary',
	height: 'auto',
	minHeight: '44px',
	class: 'text-wrap ma-2',
} as const

export const config = {
	banner: {
		rounded: true,
		elevation: 2,
		class: 'pa-8',
	},
	closeBtn: {
		icon: true,
		variant: 'text',
		width: '32px',
		height: '32px',
		class: 'ml-4',
	},
	backBtn: {
		icon: true,
		variant: 'text',
		width: '32px',
		height: '32px',
		class: 'ml-4',
	},
	customizeBtn: {
		...commonBtnConfig,
		variant: 'outlined',
	},
	rejectBtn: {
		...commonBtnConfig,
	},
	acceptBtn: {
		...commonBtnConfig,
	},
} as const
