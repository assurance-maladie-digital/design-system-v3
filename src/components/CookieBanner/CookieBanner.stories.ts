import { type Meta, type StoryObj } from '@storybook/vue3'
import CookieBanner from './CookieBanner.vue'
import { fn } from '@storybook/test'
import { VBtn } from 'vuetify/components'
import { ref } from 'vue'
import { watch } from 'fs'

const meta = {
	title: 'Composants/Feedback/CookieBanner',
	component: CookieBanner,
	parameters: {
		layout: 'fullscreen',
	},
} satisfies Meta<typeof CookieBanner>

export default meta

type Story = StoryObj<typeof meta>

export const Default: Story = {
	argTypes: {
		cookiesRoute: { control: { type: 'text' } },
		onAccept: { action: 'accept' },
		onReject: { action: 'reject' },
		onCustomize: { action: 'customize' },
	},
	args: {
		cookiesRoute: '/cookie',
		onAccept: fn(),
		onReject: fn(),
		onCustomize: fn(),
	},

	decorators: [
		() => ({
			template: '<div style="height: 500px;"><story /></div>',
		}),
	],

	render: (args) => {
		return {
			components: { CookieBanner, VBtn },
			setup() {
				watch(args.modelValue, (value) => {
					open.value = value
				}, { immediate: true })
				const open = ref(true)
				return { args, open }
			},
			template: `
			<div style="height: 500px; display: flex; align-items: center; justify-content: center;">
				<VBtn @click="open = true" v-if="!open">Reset</VBtn>
				<CookieBanner v-bind="args" v-model="open" />
			</div>
			`,
		}
	},
}

export const WithoutCookiesRoute: Story = {
	args: {
		onAccept: fn(),
		onReject: fn(),
		onCustomize: fn(),
	},
	decorators: [
		() => ({
			template: '<div style="height: 500px;"><story /></div>',
		}),
	],
}

export const DescriptionSlot: Story = {
	args: {
		onAccept: fn(),
		onReject: fn(),
		onCustomize: fn(),
	},
	render: (args) => {
		return {
			components: { CookieBanner },
			setup() {
				return { args }
			},
			template: `
			<div style="height: 500px;">
				<CookieBanner v-bind="args">
					<p><b>Custom</b> description</p>
				</CookieBanner>
			</div>
			`,
		}
	},
}

export const Customization: Story = {
	args: {
		onAccept: fn(),
		onReject: fn(),
		onCustomize: fn(),
		vuetifyOptions: {
			sheet: {
				color: '#ced9eb',
			},
			customizeBtn: {
				text: true,
				outlined: false,
			},
			rejectBtn: {
				outlined: true,
			},
			acceptBtn: {
				outlined: true,
			},
		},
	},
	render: (args) => {
		return {
			components: { CookieBanner },
			setup() {
				return { args }
			},
			template: `
			<div style="height: 500px;">
				<CookieBanner v-bind="args" />
			</div>
			`,
		}
	},
}
