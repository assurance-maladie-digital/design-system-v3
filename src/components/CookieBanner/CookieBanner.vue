<script setup lang="ts">
	import type { CustomizableOptions } from '@/composables/useCustomizableOptions'
	import useCustomizableOptions from '@/composables/useCustomizableOptions'
	import { mdiClose, mdiArrowULeftBottom } from '@mdi/js'
	import { computed, ref } from 'vue'
	import { useDisplay } from 'vuetify'
	import type { CookiesItems } from '../CookiesSelection/types'
	import { config } from './config'
	import { locales } from './locales'
	import CookiesSelection from '../CookiesSelection/CookiesSelection.vue'
	import { VBanner } from 'vuetify/components'

	const props = defineProps<CustomizableOptions & {
		items?: CookiesItems
	}>()

	const options = useCustomizableOptions(config, props)

	const emits = defineEmits(['reject', 'accept', 'customize', 'submit'])

	const active = defineModel({
		type: Boolean,
		default: true,
	})

	const showCookiesSelection = ref(false)

	const display = useDisplay()
	const btnWidth = computed(() => {
		return display.smAndDown.value ? '100%' : 'auto'
	})

	function reject(): void {
		active.value = false
		emits('reject')
	}

	function accept(): void {
		active.value = false
		emits('accept')
	}

	function customize(): void {
		if (props.items) {
			showCookiesSelection.value = true
		}
		emits('customize')
	}

	function personalizeCookies(e: Record<string, unknown>) {
		emits('submit', e)
		showCookiesSelection.value = false
		active.value = false
	}
</script>

<template>
	<Teleport
		v-if="active"
		to="body"
	>
		<VSheet
			v-bind="options.banner"
			:aria-label="locales.label"
			class="vd-cookie-banner"
		>
			<div>
				<div class="d-flex align-start flex-nowrap pa-0 mb-6">
					<h2 class="text-h5 font-weight-bold">
						{{ locales.title }}
					</h2>

					<VSpacer v-bind="options.spacer" />

					<VBtn
						v-if="showCookiesSelection"
						v-bind="options.backBtn"
						:aria-label="locales.backBtn"
						@click="showCookiesSelection = false"
					>
						<VIcon v-bind="options.icon">
							{{ mdiArrowULeftBottom }}
						</VIcon>
					</VBtn>
					<VBtn
						v-else
						v-bind="options.closeBtn"
						:aria-label="locales.closeBtn"
						@click="reject"
					>
						<VIcon v-bind="options.icon">
							{{ mdiClose }}
						</VIcon>
					</VBtn>
				</div>
				<div class="vd-cookie-banner-content">
					<Transition name="height">
						<div v-if="showCookiesSelection && items">
							<CookiesSelection
								:items
								@submit="personalizeCookies"
							>
								<template
									v-for="(_, slotName) in $slots"
									#[slotName]="slotProps"
								>
									<slot
										:name="slotName"
										v-bind="slotProps || {}"
									/>
								</template>
							</CookiesSelection>
						</div>
						<div v-else>
							<slot>
								<p>
									{{ locales.description }}
								</p>
							</slot>
	
							<div
								class="vd-cookie-banner-action-ctn d-flex align-center flex-wrap max-width-none mt-6 mb-n2 mx-n2"
							>
								<VSpacer v-bind="options.actionsSpacer" />
	
								<VBtn
									v-bind="options.customizeBtn"
									data-test-id="customize"
									:width="btnWidth"
									@click="customize"
								>
									{{ locales.customizeBtnText }}
								</VBtn>
	
								<VBtn
									v-bind="options.rejectBtn"
									data-test-id="reject"
									:width="btnWidth"
									@click="reject"
								>
									{{ locales.rejectBtnText }}
								</VBtn>
	
								<VBtn
									v-bind="options.acceptBtn"
									data-test-id="accept"
									:width="btnWidth"
									@click="accept"
								>
									{{ locales.acceptBtnText }}
								</VBtn>
							</div>
						</div>
					</Transition>
				</div>
			</div>
		</VSheet>
	</Teleport>
</template>

<style lang="scss" scoped>

.vd-cookie-banner {
	position: fixed;
	left: 50%;
	bottom: 0;
	transform: translateX(-50%);
	z-index: 20;
	max-height: calc(100dvh - 64px);
	width: calc(100% - 64px);
	max-width: 1200px;
	margin-bottom: 32px;
}

.vd-cookie-banner-content {
	overflow-y: auto;
}

.vd-cookie-banner-action-ctn .v-btn {
	flex: 1 1 auto;
}

.v-btn--icon {
	color: rgba(0, 0, 0, 0.54);
	position: absolute;
	right: 24px;
}

.height-enter-active, .height-leave-active {
	box-sizing: border-box;
	transition: height 2s ease;
	interpolate-size: allow-keywords;
	overflow: hidden;
	background-color: white;
}

.height-enter-active {
	transition: height 0.5s ease;
}

.height-leave-active {
	transition: height 0.1s ease;
}

.height-enter-from, .height-leave-to {
	height: 0;
}

.height-enter-to, .height-leave-from {
	height: auto;
}

</style>
