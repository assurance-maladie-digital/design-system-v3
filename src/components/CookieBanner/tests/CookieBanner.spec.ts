import { mount } from '@vue/test-utils'
import { describe, it, expect } from 'vitest'
import CookieBanner from '../CookieBanner.vue'
import { vuetify } from '@tests/unit/setup'
describe('CookieBanner', () => {
	it('renders correctly', () => {
		const wrapper = mount(CookieBanner, {
			global: {
				plugins: [vuetify],
			},
		})

		expect(wrapper.html()).toMatchSnapshot()
	})

	it('renders correctly on small screens', async () => {
		// @ts-expect-error  - Property 'happyDOM' does not exist on type 'Window & typeof globalThis'.
		window.happyDOM.setInnerWidth(600)

		const wrapper = mount(CookieBanner, {
			global: {
				plugins: [vuetify],
				stubs: {
					Teleport: true,
				},
			},
		})
		await wrapper.vm.$nextTick()

		expect(wrapper.find('[data-test-id="customize"]').attributes('style')).toContain('100%')
	})

	it('emit a reject event when the reject btn is clicked', async () => {
		const wrapper = mount(CookieBanner, {
			global: {
				plugins: [vuetify],
				stubs: {
					Teleport: true,
				},
			},
		})

		await wrapper.find('[data-test-id="reject"]').trigger('click')

		expect(wrapper.emitted()).toHaveProperty('reject')
	})

	it('emit a accept event when the accept btn is clicked', async () => {
		const wrapper = mount(CookieBanner, {
			global: {
				plugins: [vuetify],
				stubs: {
					Teleport: true,
				},
			},
		})

		await wrapper.find('[data-test-id="accept"]').trigger('click')

		expect(wrapper.emitted()).toHaveProperty('accept')
	})

	it('does not close the dialog when the customize button is clicked and do not show the cookie form', async () => {
		const wrapper = mount(CookieBanner, {
			global: {
				plugins: [vuetify],
				stubs: {
					Teleport: true,
				},
			},
		})

		expect(wrapper.find('.vd-cookie-banner').exists()).toBe(true)
		await wrapper.find('[data-test-id="customize"]').trigger('click')
		await wrapper.vm.$nextTick()
		expect(wrapper.find('.vd-cookie-banner').exists()).toBe(true)
		expect(wrapper.find('.vd-cookies-card').exists()).toBe(false)
		expect(wrapper.emitted()).toHaveProperty('customize')
	})

	it('does not close the dialog when the customize button is clicked and show the cookie form', async () => {
		const wrapper = mount(CookieBanner, {
			props: {
				items: {
					essentials: [
						{
							name: 'session',
							description: 'Sauvegarde la session pour rester connecté.',
							conservation: '20 heures',
						},
						{
							name: 'cookie_policy',
							description: 'Sauvegarde les préférences de cookies.',
							conservation: '1 an',
						},
					],
					functional: [
						{
							name: 'contrast',
							description: 'Sauvegarde la personnalisation de l’affichage.',
							conservation: '1 an',
						},
					],
					analytics: [
						{
							name: 'user_id',
							description: 'Sauvegarde l’identifiant unique de visiteur.',
							conservation: '6 mois',
						},
					],
				},
			},
			global: {
				plugins: [vuetify],
				stubs: {
					Teleport: true,
				},
			},
		})

		expect(wrapper.find('.vd-cookie-banner').exists()).toBe(true)
		await wrapper.find('[data-test-id="customize"]').trigger('click')
		await wrapper.vm.$nextTick()
		expect(wrapper.find('.vd-cookie-banner').exists()).toBe(true)
		expect(wrapper.find('.vd-cookies-card').exists()).toBe(true)
		expect(wrapper.find('.vd-cookies-card').html()).toMatchSnapshot()
		expect(wrapper.emitted()).toHaveProperty('customize')
	})
})
