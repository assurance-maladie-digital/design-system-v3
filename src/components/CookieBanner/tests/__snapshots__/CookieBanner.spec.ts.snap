// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CookieBanner > renders correctly 1`] = `
"<!--teleport start-->
<!--teleport end-->"
`;
